### ISSUE TITLE
Implement `numpy.random.choice` equivalent

### ISSUE THREAD

**Issue by LeviViana** [pytorch/pytorch#16897]
## 🚀 Feature
Implement `numpy.random.choice` equivalent.

## Motivation
In some cases, it is useful to get random samples from a torch Tensor efficiently. For instance, during the training of some Region-Based Detectors, it is necessary to control the proportion of positive and negative regions of interest (RoIs) over mini-batches. 

Here is a workaround adopted in maskrcnn-benchmark: https://github.com/facebookresearch/maskrcnn-benchmark/blob/master/maskrcnn_benchmark/modeling/balanced_positive_negative_sampler.py#L49-L50, which can be inefficient if `positive.numel()` is big and `num_pos` is small, for instance.

## Pitch

Implement `torch.random.choice` to have an equivalent behavior to the numpy implementation.

**Comment by LeviViana** [comment]
BTW, it looks like the numpy implementation for doing that kind of sampling follows the same workaround as maskrcnn-benchmark: https://github.com/numpy/numpy/blob/76a76c78d1b049126153e81b0a9d137fa3e4947b/numpy/random/mtrand/mtrand.pyx#L1194

**Comment by zou3519** [comment]
Sounds like we could implement a reservoir sampling algorithm: https://en.wikipedia.org/wiki/Reservoir_sampling

**Comment by LeviViana** [comment]
I think I found a simple solution: use `torch.randint` for the indices.

For instance, if we try to sample 10 elements from a Tensor of first dimension size `10**6`, then we get:

- Usual workaround: `python -m timeit --setup="import torch;x=torch.arange(10**6)" "torch.randperm(x.numel())[:10]"`
- The torch.randint trick: `python -m timeit --setup="import torch;x=torch.arange(10**6)" "x[torch.randint(0, x.size(0), (10,))]"`

**Comment by LeviViana** [comment]
I've created a repo for this subject here: https://github.com/LeviViana/torch_sampling. I'm still working on the CUDA implementation. I'd be happy to read your comments about this implementation.

**Comment by ja3067** [comment]
@LeviViana I just ran into this lack of functionality, and I'd be happy to help with the CUDA implementation or the bindings. Is there anything that remains to be done?

**Comment by LeviViana** [comment]
@ja3067 Thanks for proposing your help. Yes, your help is very welcome. I don't have anything running properly on CUDA for the moment. So, what do you think of developing this functionality on [this repo](https://github.com/LeviViana/torch_sampling)? 

Only the uniform distribution supported for the moment. But sure, it would be nice to feed the distribution to the function, just as `numpy.random.choice`.

**Comment by LeviViana** [comment]
A first version of a full-featured `numpy.random.choice` equivalent for PyTorch is now available [here](https://github.com/LeviViana/torch_sampling) (working on PyTorch 1.0.0). It includes CPU and CUDA implementations of:

- Uniform Random Sampling WITH Replacement (via `torch::randint`)
- Uniform Random Sampling WITHOUT Replacement (via reservoir sampling)
- Weighted Random Sampling WITH Replacement (via inverse transform sampling)
- Weighted Random Sampling WITHOUT Replacement (via [this method](https://utopia.duth.gr/~pefraimi/research/data/2007EncOfAlg.pdf))

@ja3067 @zou3519 @ailzhang @gchanan

**Comment by gchanan** [comment]
neat!  Note this would take some work to get in to PyTorch proper, namely specifying as a native function, swapping `torch` namespace to `at` namespace, etc.

**Comment by LeviViana** [comment]
Thanks for your reply. I'll be making these changes and sending a PR soon.

**Comment by awarebayes** [comment]
Any progress?

**Comment by LeviViana** [comment]
The PR is currently being reviewed.

**Comment by superRookie007** [comment]
Is this feature going to be released soon?
BTW, the implementation of choice in numpy has a very annoying [issue](https://github.com/numpy/numpy/pull/6131), where it would encounter numerical problems.

**Comment by gchanan** [comment]
The PR above hasn't made it in yet, so probably not soon.

Thanks for the pointer to the NumPy issue.

**Comment by OleguerCanal** [comment]
Any news?

**Comment by namanjaswani27** [comment]
If you want to use` np.random.choice` equivalent in PyTorch to increase variation in output text of a Text Generation model, try using `torch.multinomial` [[link]](https://pytorch.org/docs/stable/generated/torch.multinomial.html)

**Comment by nihal-j** [comment]
Any updates to this?

**Comment by zou3519** [comment]
Increasing priority due to activity

**Comment by BehzadShayegh** [comment]
Any news?

**Comment by amitport** [comment]
If anybody is interested I found that for the uniform sampling without replacement case, the following works fairly well (kind of slow if the input vecor dim ~> 10mil)
```python
# given a tensor x this is equivalent to 'x[numpy.random.choice(n, size=k, replace=False)]`
coordinate_score = torch.empty_like(x).uniform_()
indices = torch.topk(coordinate_score, k, sorted=False).indices
return x[indices]
```

**Comment by apple2373** [comment]
I also want this function

**Comment by kmfrick** [comment]
This might be helpful in the C++ API, where indexing with `.index()` is slow (see issue #20263)

**Comment by franchesoni** [comment]
Nudging here

**Comment by Luvideria** [comment]
Is there some actual roadblocks that prevent this feature from being integrated?

**Comment by kmfrick** [comment]
[This](https://github.com/pytorch/pytorch/pull/18624) is the relevant PR. It was being compared to `torch._multinomial_alias_draw`, but that function has been killed, so I'm honestly not sure what's going on here.

**Comment by yoandgo42** [comment]
also ran into this problem

**Comment by vak** [comment]
marked as a high priority. are there any update? 
it would be really cool to have this **fundamental** feature.

**Comment by brucelightyear** [comment]
Up !

**Comment by ducha-aiki** [comment]
up

**Comment by Evidlo** [comment]
If you need weighted sampling, this is what I came up with:

``` python
def inverse_sampling(grid, dist, size):
    """Do inverse transform sampling to sample from a grid with given probabilities
    https://en.wikipedia.org/wiki/Inverse_transform_sampling

    Args:
        grid (tensor): Random are samples pulled from this 1D tensor
        dist (tensor): Probability of each item in the grid.  Should sum to 1
        size (tuple): shape of sampled output

    Returns:
        sampled (tensor): sampled values from `grid` with shape `size`
    """
    grid = t.tensor(grid, device='cuda')
    dist = t.tensor(dist, device='cuda')
    dist_cum = t.cumsum(dist, 0)
    rand_ind = t.searchsorted(dist_cum, t.rand(size=size, device='cuda'))
    return grid[rand_ind]
```

**Comment by nkalpakis21** [comment]
i see this has been open for awhile. is there trouble getting a dev to fix this issue? what is blocking this for so long

**Comment by Flecart** [comment]
News?

**Comment by clay-arras** [comment]
+1 for this feature, would be really useful

**Comment by amitport** [comment]
@clay-arras `torch.distributions.categorical.Categorical` does not support sampling without replacement while `numpy.random.choice` does
